BYOIDP GA (#6021)
mendix/docs: content/en/docs/developerportal/control-center/set-up-sso-byoidp.md

* When you add a domain to your company account, it is automatically added to the active IdP configuration. 
* Users with domains which are not part of your company are unaffected. They still have access based on the way they normally sign in to Mendix.
* Routing of the login process to your IdP is based on the user's email domain (as entered in the Mendix login screen). If you have invited users from other organisations to your projects in Mendix (or contractors), enabling the BYOIDP will not affect the way those users are authenticated since they're associated with a different email domain.
* When BYOIDP is used, a session at Mendix will be valid for 1 hour. After the session is expired, Mendix will request a new ID_token from your IdP. If the user still has a session at your IdP, the token will be issued without any user input and the platform user continues to have access to the Mendix Platform. The effect of this mechanism is that users have access to the Mendix Platform as long as the session at your IdP is valid.

> * Applications built with Mendix SSO can delegate authentication to your IdP. Such applications can use an 'idp_hint' to pre-select the IdP and bypass the Mendix login screen. 

Author: We don't have any documentation on 'idp_hint'. Can we explain more here, or put the information in the Mendix SSO document?

Reviewer: Let's remove the idp_hint statement altogether, since we're not really promoting Mendix SSO anymore for production purposes.

Author: OK
```suggestion
```

#### 1.2.3 Technical Integration

BYOIDP SSO integrates with the Mendix Platform using the following techniques:

* BYOIDP uses the IdP's OIDC well-known/discovery endpoint to retrieve the URLs for the authorization endpoint, the token endpoint, and the jwks endpoint.

> * The user's email address is used to associate a user's existing account within Mendix with the user's account at your IdP. This means that any existing Mendix account is linked to their IdP account for authentication, rather than a new Mendix account being created for them.

Author: Is this, in fact, the user's email address.
I think it is the user's **Username** (although currently I don't think you can change these independently - you used to be able to!).

Reviewer: technically these may be diferent things but from auser perspective your email address is used as username and they are the same, so...

Author: I'll leave it as is, then.

* Mendix Platform APIs which require a PAT, can use one which is created by a platform user. Alternatively, you can set up a service account to consume Mendix platform APIs in one of the following ways:

    * Use a personal account as if it were a service account
    * Create a service account in the company IdP
    * Create a service account on an email domain that is not federated with BYOIDP

* When the Company Admin activates the 'BYOIDP configuration', Mendix scrambles the Mendix passwords for all impacted users. De-activation of the feature doesn't roll-back those changes. If you deactivate BYOIDP SSO, users have to reset their Mendix password before being able to login with their Mendix account.
* Your conditional access policies in AzureAD may block Studio Pro logins for versions of Mendix below 9.18, as these make use of an embedded browser. If you are using Microsoft's Intune for MDM/MAM and versions of Mendix below 9.18, you may not want to activate BYOIDP for this reason.

> * You cannot associate multiple Mendix accounts (e.g. an Admin account and a regular account) with a single identity in your IdP. If such a mapping is needed it needs to be catered for in your IdP.

Author: Since we don't support this, I'm not sure how any IdP-based mapping could work.
Can we explain this a bit better - or perhaps remove the last sentence?

Reviewer: yes, let's remove the second sentence!

Author: ```suggestion
* You cannot associate multiple Mendix accounts (e.g. an Admin account and a regular account) with a single identity in your IdP.
```

* **Convenience**. Platform users have convenience of SSO and don’t have to manage credentials for the Mendix platform.

### 1.2 Features

#### 1.2.1 User Experience

When end-user who has BYOIDP SSO enabled attempts to sign in to the Mendix Platform or a Mendix SSO supported app, they enter their email address as a **Username**. The sign in process will recognize that the email domain is configured for SSO and remove the password field, showing a button labeled **Sign in with SSO**. 

#### 1.2.2 General Features

BYOIDP SSO has the following features:

* Mendix platform services and Studio Pro can delegate authentication to the customer’s IdP.
* Authentication will be delegated for any user that has an email address for which the email-domain is associated with your company. This includes service accounts (i.e. non-personal accounts ) that may have been created on the Mendix platform (e.g. “mendix@mycompany.com”). 
* When you add a domain to your company account, it is automatically added to the active IdP configuration. 
* External users are unaffected; they would still have access based on Mendix credentials or via BYOIDP-SSO to their ‘home’ IdP.

> * Routing of the login process to the customer’s IdP is based on the user’s email domain (as entered in the Mendix login screen). If you have invited users from other organisations to your projects in Mendix (or contractors), enabling the BYOIDP will not affect the way those users are authenticated since they’re associated with a different email domain.

Author: This seems to just repeat the previous two bullet points:
```suggestion
```

tags: ["BYOIDP", "IdP", "Bring Your Own IdP", "Microsoft Azure", "SSO", "Single Sign-on"]
---

## 1 Introduction

> The Mendix Cloud contains an Identity Provider (IdP) that allows users to sign in to:

Reviewer: I'd say "Mendix platform" rather than "Mendix Cloud".

Author: ```suggestion
The Mendix Platform contains an Identity Provider (IdP) that allows users to sign in to:
```

* **Convenience**. Platform users have the convenience of SSO and don't have to manage credentials for the Mendix Platform.

### 1.2 Features

#### 1.2.1 User Experience

When a user who has BYOIDP SSO enabled attempts to sign in to the Mendix Platform, Studio Pro, or a Mendix SSO supported app, they enter their email address as a **Username**. The sign in process will recognize that the email domain is configured for SSO and remove the password field, showing a button labeled **Sign in with SSO**. 

#### 1.2.2 General Features

BYOIDP SSO has the following features:

* Mendix platform services, Studio Pro, and Mendix apps using Mendix SSO can delegate authentication to your IdP.
* Authentication will be delegated for any user that has an email address where the email-domain is associated with your company. This includes service accounts (for example non-personal accounts used for consuming APIs) that may have been created on the Mendix Platform. 
* When you add a domain to your company account, it is automatically added to the active IdP configuration. 

> * Users with domains which are not part of your company are unaffected. They still have access based on the way they normally sign in to Mendix.

Reviewer: 'that' instead of 'which' maybe?
I'm assuming that=dat and which=wat in Dutch ;-)

Reviewer: Can we use the phrase 'external users' as well. I think other pages on docs.mendix.com use that same terminology. I think that contributes to clarity.

Author: OK - I don't like external users as it seems a bit imprecise to me, but if we use it elsewhere I'll change it!
_Which_ and _that_ are interchangeable here - I tend to use "which" but Microsoft Word grammar checker prefers "that" so I'll change it.
```suggestion
* External users (with domains that are not part of your company) are unaffected. They still have access based on the way they normally sign in to Mendix.
```

* Routing of the login process to your IdP is based on the user's email domain (as entered in the Mendix login screen). If you have invited users from other organisations to your projects in Mendix (or contractors), enabling the BYOIDP will not affect the way those users are authenticated since they're associated with a different email domain.
* When BYOIDP is used, a session at Mendix will be valid for 1 hour. After the session is expired, Mendix will request a new ID_token from your IdP. If the user still has a session at your IdP, the token will be issued without any user input and the platform user continues to have access to the Mendix Platform. The effect of this mechanism is that users have access to the Mendix Platform as long as the session at your IdP is valid.
* Applications built with Mendix SSO can delegate authentication to your IdP. Such applications can use an 'idp_hint' to pre-select the IdP and bypass the Mendix login screen. 

#### 1.2.3 Technical Integration

BYOIDP SSO integrates with the Mendix Platform using the following techniques:

> * BYOIDP uses the IdP's OIDC well-known/discovery endpoint to retrieve the URLs for the authorization endpoint, the token endpoint, and the jwks endpoint.

Author: I'll use OIDC throughout the document, and just mention OpenID Connect once, here.
```suggestion
* BYOIDP uses the IdP's OIDC (OpenID Connect) well-known/discovery endpoint to retrieve the URLs for the authorization endpoint, the token endpoint, and the jwks endpoint.
```

* The user's email address is used to associate a user's existing account within Mendix with the user's account at your IdP. This means that any existing Mendix account is linked to their IdP account for authentication, rather than a new Mendix account being created for them.

    This assumes that the IdP returns an email address to Mendix during SSO which the user previously used to sign-up and login to Mendix. If the email address that is returned to Mendix is not recognized, then the user will be offered the sign-up option to enable them to create a new account.

> * The authentication/authorization request to your IdP typically requests the 'openid' and 'profile' scopes as defined by OIDC. The request does not include a scope to request authorization for the user to use the platform or part of the platform. Your IdP may apply coarse grained access rules based on the client_id for the Mendix Platform.

Author: I don't find it very clear what you mean by applying course-grained access rules when you also say that you can't request any scopes?
Can we make this clearer?

Reviewer: Yes, this should be more clear. How about:

"The request to your IdP is technically an authentication request, which on protocol level means that only the 'openid' and 'profile' scope values are requested, as defined by OIDC. The request does not explicitly ask for authorisation for specific platform roles such as 'developer', 'Mendix Administrator', 'technical contact'; i.e. on protocol-level no scope values that resemble such roles will be included in the request. Note that your IdP, however, may apply coarse grained access rules based on the client_id for the Mendix Platform to deny access to the Mendix platform for certain groups of employees."

See also my blog https://paper.dropbox.com/doc/Yes-bring-your-own-IdP--B4Gt3gbHddQoj_jiVx4jTj9eAg-TgA09M74Be8BEHZ6DL8sW ;-)

Author: I've simplified your proposal a bit, hope it still makes sense.
```suggestion
* BYOIDP SSO makes an authentication request to your IdP which means that only the 'openid' and 'profile' scope values are requested, as defined by OIDC. The request does not explicitly ask for authorisation for specific platform roles such as 'developer', 'Mendix Administrator', 'technical contact'. You can set up your IdP, however, to apply coarse-grained access rules based on the client_id for the Mendix platform to deny access to the Mendix platform for certain groups of employees.
```

Reviewer: yes, agreed

* Mendix provides support for two client authentication methods: client_secret_post (client credentials in the payload) or client_secret_basic (basic authentication credentials in http header). If the IdP supports both methods, client_secret_post is used.
* Mendix includes the `login_hint` parameter in requests to your IdP This allows the IdP to pre-populate the login screen with the user's email address, which gives a better user experience. Your IdP may choose to ignore the hint. After receiving a positive response, Mendix does not do any validation if the logged-in user matches the login_hint.
* Whether or not end-users signing in to the Mendix Platform have to use 2FA does not change the [Two-Factor Authentication](/developerportal/deploy/two-factor-authentication/) which protects sensitive activities on Mendix Cloud nodes. This remains in place and works independently of BYOIDP SSO.

> ### 1.3 Limitations

Reviewer: I think it's good to also add the following limitations:

"When a user gets a new email address in the customer’s IDP, the logic within Mendix may not recognise it as the same account and will ask the user to do a new signup.

Some customer’s have problems with processing or ignoring the login_hint, which is part of the OIDC protocol. Mendix currently cannot leave out the login_hint from the authentication requests.

Author: ```suggestion
### 1.3 Limitations

BYOIDP SSO has the following limitations.

* If the end-user's email address is changed in your IdP, Mendix may not recognize it as the same account and will ask the end-user to set up a new Mendix account.
* `login_hint` is not optional and is always sent as part of authentication requests to the IdP.
```

Reviewer: ok

> 

Author: ```suggestion
```

> * Your IdP must support OIDC (Microsoft Azure AD is one example of such an IdP).

Reviewer: It's not really a limitation.
We can do two things:
- say that BYOIDP requires the IdP to support OIDC (w're saying that in technical integration section altready)
- say that we don't support SAML protocol for BYOIDP-SSO. (limitations)

I'd suggest to do both.

Author: ```suggestion
* BYOIDP SSO only supports OIDC and does not support other protocols such as SAML.
```

> * Your Mendix app must be built using Mendix version 7.23 or above.

Reviewer: It's not so much about the apps, but it's a about the version of StudioPro (which has implicatiosn for the versions for the apps).

That's why in the dropbox paper I said: "Only Studio Pro versions 7.18 and above support SSO.  Studio Pro 7.23 LTS can be used to edit apps built with Studio Pro 6.0.0 - 7.23. Therefore, before enabling BYOIDP, customers should upgrade their apps from before 6 LTS to at least 7.23 (LTS version)."

Makes more sense now?

Author: I disagree, the version of Studio Pro supports a specific version of Mendix (not just Studio Pro but the runtime which interprets the deployment package and whether it will work with modules out of the Marketplace which might have been built with a later version of Mendix).

I don't understand your statement from Dropbox, it just seems a very complicated way of saying that the Mendix app must be 7.23 or above. I don't know what additional information your statement gives.

Reviewer: Some StudioPro versions allow to open an app that's still on an older version.
That's what the dripbox paper is trying to say.

With Mendix 10, Mendix7 goes out of support , so this becomes less relevant.

Author: **All** Studio Pro versions allow you to open an app which is older¹. You just can't open an app which has a higher version number. Therefore I don't think we need to complicate things here.

¹ For some combinations there may be breaking changes which means it won't run - but it will happily upgrade it for you.

> * Once BYOIDP is activated, direct access to [Team Server](/developerportal/collaborate/team-server/) is no longer possible using a username and password. To access code repositories from a pipeline, you need to use a Personal Access Token (PAT).

Reviewer: This needs to be removed. This is the limitation we are removing by introducing PAT for SVN (!!)

This means the PR for PAT-based access to SVN must be published first.

Author: That is what this text says. That the customer has to use a PAT once BYOIDP is activated. It doesn't say they can't access it at all.

Reviewer: You're right. My bad.

* Mendix Platform APIs which require a PAT can use one which is created by a platform user. You cannot directly set up service accounts within Mendix once BYOIDP SSO is activated. You can set up a service-like account to consume Mendix platform APIs in one of the following ways:

    * Use a personal account as if it were a service account
    * Create a service account in the company IdP
    * Create a service account on an email domain that is not federated with BYOIDP

* When the Company Admin activates the 'BYOIDP configuration', Mendix scrambles the Mendix passwords for all impacted users. De-activation of the feature doesn't roll-back those changes. If you deactivate BYOIDP SSO, users have to reset their Mendix password before being able to login with their Mendix account.

> * Your conditional access policies in AzureAD may block Studio Pro logins for versions of Mendix below 9.18, as these make use of an embedded browser. If you are using Microsoft's Intune for MDM/MAM and versions of Mendix below 9.18, you may not want to activate BYOIDP for this reason.

Reviewer: This is about versions of Studio Pro again. this can be more clear.

So maybe combine with that previous point?

Author: This is more debatable, but I still think that customers need to be clear that using a specific Mendix version has more impact than just what features are in Studio Pro.

Reviewer: As per my clarification on similar comment.

My understanding is that a version of StudioPro support allows opening projects from multiple runtime versions. The version of StudioPro is relevant for SSO.

* You cannot associate multiple Mendix accounts (e.g. an Admin account and a regular account) with a single identity in your IdP. If such a mapping is needed it needs to be catered for in your IdP.

## 2 Prerequisites

To set up an IdP configuration for the Mendix Platform and your Mendix app, you will need the following:

> * A subscription to an OpenID Connect compliant IdP. Validate that your IdP supports OpenID Connect

Reviewer: Is it clear to the reader that OpenID Connect  = OIDC?

Author: Is it the same 😀- I'll be consistent through the doc and always use OIDC (I've made the connection the first time I mention it).
```suggestion
* A subscription to an OIDC compliant IdP. Validate that your IdP supports OIDC.
```

>     * If you are using Azure AD, Okta, Auth0, or ADFS you will be OK

Reviewer: or Ping Identity https://www.pingidentity.com/en.html

Author: ```suggestion
    * If you are using Azure AD, Okta, Auth0, [Ping Identity](https://www.pingidentity.com), or ADFS you will be OK
```

### 1.2 Features

#### 1.2.1 User Experience

When a user who has BYOIDP SSO enabled attempts to sign in to the Mendix Platform, Studio Pro, or a Mendix SSO supported app, they enter their email address as a **Username**. The sign in process will recognize that the email domain is configured for SSO and remove the password field, showing a button labeled **Sign in with SSO**. 

#### 1.2.2 General Features

BYOIDP SSO has the following features:

* Mendix platform services, Studio Pro, and Mendix apps using Mendix SSO can delegate authentication to your IdP.
* Authentication will be delegated for any user that has an email address where the email-domain is associated with your company. This includes service accounts (for example non-personal accounts used for consuming APIs) that may have been created on the Mendix Platform. 
* When you add a domain to your company account, it is automatically added to the active IdP configuration. 
* External users (with domains that are not part of your company) are unaffected. They still have access based on the way they normally sign in to Mendix.

> * Routing of the login process to your IdP is based on the user's email domain (as entered in the Mendix login screen). If you have invited users from other organisations to your projects in Mendix (or contractors), enabling the BYOIDP will not affect the way those users are authenticated since they're associated with a different email domain.

Author: @JaapF 
This seems to duplicate the previous two bullet points. I think we can remove it:
```suggestion
```

Author: @JaapF 
Only this suggestion outstanding - then we can publish (if you are also happy with the last change I made to #6029 in response to your suggestion)

* **Configuration name** – a name for the IdP setup in the Single Sign-On configuration for your own reference.

    {{< figure src="/attachments/developerportal/control-center/set-up-sso-byoidp/customer-idp-wizard-page-1.png" >}}

* **Redirect URL** – the callback URL to the Developer Portal that your IdP needs to send the authenticated user to.

> * **OpenID Connect well-known endpoint URL** – the endpoint at your IdP from which the Mendix Platform can retrieve the configuration metadata, including all necessary endpoints and public key location information.

Author: If we want to use OIDC, then perhaps we should update the UX to use OIDC as well?

Reviewer: Let's not worry. In section 1.2.3. it is maed clear that OIDC means OpenId Connect.

>     If your IdP supports multiple protocols, make sure you enter the OpenID Connect endpoint.

Author: ```suggestion
    If your IdP supports multiple protocols, make sure you enter the OIDC endpoint.
```

* **Client ID** – the ID of the Developer Portal registration in your IdP.
* **Client secret** – the password or secret of the Developer Portal registration in your IdP. Enter this once. After saving your configuration it will no longer be shown to you. See [Changing the Client Secret](#client-secret), below, for information about changing this value once your configuration is active.

> * **Scopes** – selecting a scope to configure the data Mendix is allowed to read from your IdP. Mendix uses this data to map the user's identity in your IdP environment with a corresponding identity in the Developer Portal. The scope `OpenID` is required. In some cases, depending on your IdP, other scopes are necessary to fully map the user's identity.

Reviewer: I suggest to add:

Typically the scopes "profile" and "email" are needed to get the user's email address and name, which are required for SSO with the Mendix platform.

Author: ```suggestion
* **Scopes** – selecting a scope to configure the data Mendix is allowed to read from your IdP. Mendix uses this data to map the user's identity in your IdP environment with a corresponding identity in the Developer Portal. The scope `OpenID` is required. Typically the scopes "profile" and "email" are also needed to get the end-user's email address and name, which are required for SSO to fully map the end-user's identity. Your IdP may provide additional scopes you can use.
```

    {{< figure src="/attachments/developerportal/control-center/set-up-sso-byoidp/customer-idp-wizard-page-2.png" >}}

### 3.2 Testing Your Configuration

> Once you have configured the endpoint and the scope, you are ready to perform the first test of your configuration. This test will perform a round trip from the Mendix Platform to your IdP and back to the platform. The test will be performed in a new tab page of your browser so your browser must allow popup windows.

Reviewer: "scopes" not "scope"

Author: ```suggestion
Once you have configured the endpoint and the scopes, you are ready to perform the first test of your configuration. This test will perform a round trip from the Mendix Platform to your IdP and back to the platform. The test will be performed in a new tab page of your browser so your browser must allow popup windows.
```

{{< figure src="/attachments/developerportal/control-center/set-up-sso-byoidp/customer-idp-wizard-page-4.png" >}}

## 4 Testing

When you have completed your IdP setup, you can perform a test sign-in before activating the configuration.

You can test in one of two ways:

1. Click on the test endpoint of your configuration. It will redirect you to the login page of your IdP. Enter the credentials of a user known to your IdP. If the test succeeds the Developer Portal landing page will open.
2. Go to https://login.mendix.com/ and, in the username field, enter the test email domain of your configuration exactly as printed on the overview page. The password field should disappear. Click the 'Sign in with SSO' button. This will redirect you to a login page of your IdP. Enter credentials known to your IdP. If the test succeeds the Mendix Platform home page will open.

If your test fails, see the [Troubleshooting](#troubleshooting) section for advice on where to look for issues.

## 5 Activating

Before activating your BYOIDP configuration, ensure you have read the [Considerations](#considerations) section, below.

> When you are ready, you can activate the IdP configuration from the overview page. Although the changes start to come into effect immediately, updating your users will take some time depending on the number of users in your company. Your users will start to benefit from logging into the Mendix Developer Portal with the same credentials as they use in the IdP. Any user passwords currently held in the Mendix Platform will be scrambled to prevent the users bypassing your IdP authentication.

Author: As suggested below:
```suggestion
When you are ready, you can activate the IdP configuration from the overview page. This will have an immediate effect on the interactive login process. Mendix will also scramble the platform passwords which will take some time depending on the number of users in your company. Once all the passwords are scrambled, users will no longer be able to use these for direct access to the Team Server for apps using SVN for version control.
```

> 

Reviewer: I think this may be a bit confusing. How about:

When you are ready, you can activate the IdP configuration from the overview page. This will have an immediate effect on the interactive login process. Mendix will also scramble the platform passwords, this will take some time depending on the number of users in your company. Once all passwords are scrambled, users will no longer be able to use those for direct access to Teamsever:SVN.

Author: Made a proposal above.

## 6 Deactivating

You can deactivate the IdP configuration at any time. Although the changes start to come into effect immediately, updating your users will take some time depending on the number of users in your company. Users will have to reset their Mendix password to be able to sign-in since it was scrambled upon activation.

## 7 Microsoft Azure AD

> Microsoft Azure AD is one of the most used IdP's, and supports OpenID Connect. To help you setup with Azure, follow the steps in this small tutorial.

Author: ```suggestion
Microsoft Azure AD is one of the most used IdP's, and supports OIDC. To help you setup with Azure, follow the steps in this small tutorial.
```

Once BYOIDP is activated, direct access to Team Server is no longer possible. To access code repositories from a pipeline, you need to use a [PAT](/developerportal/community-tools/mendix-profile/#pat).

Before activating BYOIDP, your developers should set up PATs for direct access to repos (for example, from CI/CD pipelines and/or Tortoise) instead of using usernames and passwords.

If developers haven't created a PAT before BYOIDP SSO is activated, they can do so later, if needed.

{{% alert color="info" %}}
Access to Team Server through other mechanisms, such as via Studio Pro or using Mendix for Private Cloud, is not affected.
{{% /alert %}}

### 8.3 Changing the Client Secret {#client-secret}

Once you have set up BYOIDP SSO, you may want to change the 'client secret' that is used by the Mendix Platform to communicate securely with the SSO.

If the client secret is still active, you can do one of the following:

* Clone the existing active IdP configuration, update the secret in there and then activate the clone. This avoids updating all the users, etc. This is the preferred option.
* Disable the IdP configuration, update the secret and then enable it again.
    

> If the client secret has expired, you can ask Mendix Support to update the client secret of your active IdP configuration.

Reviewer: "If the client secret has expired and you no longer have access to the platform to execute the above steps, you can ask Mendix Support to update the client secret of your active IdP configuration or ask them to deactivate your IdP configuration."

Reviewer: Sorry next line, says a similar thing.
I wanted to add "and you no longer have access".
If secret has expired and admin user still has a session, he could still do the necessary updates,

Author: Made a proposal below.

> If this is unsuccessful, Mendix Support will need to deactivate the active IdP configuration. This means that you (and your users) will need to reset and then use your platform password. Once you have access to the platform, you can set up your IdP configuration again.

Author: From comment above:
```suggestion
If this is unsuccessful, you can ask Mendix Support to deactivate the active IdP configuration. This means that you (and your users) without active sessions will no longer have access and will need to reset and then use your platform password. Once you have access to the platform, you can set up your IdP configuration again.
```

Reviewer: good. ok.

> If this is unsuccessful, Mendix Support will need to deactivate the active IdP configuration. This means that you (and your users) will need to reset and then use your platform password. Once you have access to the platform, you can set up your IdP configuration again.

Author: From comment above:
```suggestion
If this is unsuccessful, you can ask Mendix Support to deactivate the active IdP configuration. This means that you (and your users) without active sessions will no longer have access and will need to reset and then use your platform password. Once you have access to the platform, you can set up your IdP configuration again.
```
